x-pack/plugins/ml/public/data_frame/components/define_pivot/define_pivot_form.tsx

```diff
@@ -0,0 +1,248 @@
+  pivotSupportedAggs,
+} from '../../common';
+
+import { IndexPatternContext } from '../../common';
+
+export interface DefinePivotExposedState {
+  aggList: Label[];
+  aggs: OptionsDataElement[];
+  groupBy: Label[];
+  search: string;
+  valid: boolean;
+}
+
+const defaultSearch = '*';
+const emptySearch = '';
+
+export function getDefaultPivotState() {
+  return {
+    aggList: [] as Label[],
+    groupBy: [] as Label[],
+    search: defaultSearch,
+    valid: false,
+  } as DefinePivotExposedState;
+}
+
+interface Props {
+  overrides?: DefinePivotExposedState;
+  onChange(s: DefinePivotExposedState): void;
+}
+
+export const DefinePivotForm: SFC<Props> = React.memo(({ overrides = {}, onChange }) => {
+  const defaults = { ...getDefaultPivotState(), ...overrides };
+
+  const indexPattern = useContext(IndexPatternContext);
+
+  if (indexPattern === null) {
+    return null;
+  }
+
+  const fields = indexPattern.fields
+    .filter(field => field.aggregatable === true)
+    .map(field => ({ name: field.name, type: field.type }));
+
+  // The search filter
+  const [search, setSearch] = useState(defaults.search);
+
+  const addToSearch = (newSearch: string) => {
+    const currentDisplaySearch = search === defaultSearch ? emptySearch : search;
+    setSearch(`${currentDisplaySearch} ${newSearch}`.trim());
+  };
+
+  const searchHandler = (d: ChangeEvent<HTMLInputElement>) => {
+    const newSearch = d.currentTarget.value === emptySearch ? defaultSearch : d.currentTarget.value;
+    setSearch(newSearch);
+  };
+
+  // The list of selected group by fields
+  const [groupBy, setGroupBy] = useState(defaults.groupBy as Label[]);
+
+  const addGroupBy = (d: DropDownLabel[]) => {
+    const label: Label = d[0].label;
+    const newList = uniq([...groupBy, label]);
+    setGroupBy(newList);
+  };
+
+  const deleteGroupBy = (label: Label) => {
+    const newList = groupBy.filter(l => l !== label);
+    setGroupBy(newList);
+  };
+
+  // The list of selected aggregations
+  const [aggList, setAggList] = useState(defaults.aggList as Label[]);
+
+  const addAggregation = (d: DropDownLabel[]) => {
+    const label: Label = d[0].label;
+    const newList = uniq([...aggList, label]);
+    setAggList(newList);
+  };
+
+  const deleteAggregation = (label: Label) => {
+    const newList = aggList.filter(l => l !== label);
+    setAggList(newList);
+  };
+
+  // The available fields for group by
+  const groupByOptions: EuiComboBoxOptionProps[] = [];
+  fields.forEach(field => {
+    const o: DropDownLabel = { label: field.name };
+    groupByOptions.push(o);
+  });
+
+  // The available aggregations
+  const aggOptions: EuiComboBoxOptionProps[] = [];
+  const aggOptionsData: Dictionary<OptionsDataElement> = {};
+
+  fields.forEach(field => {
+    const o: DropDownOption = { label: field.name, options: [] };
+    pivotSupportedAggs.forEach(agg => {
+      if (
+        (agg === 'cardinality' && (field.type === 'string' || field.type === 'ip')) ||
```

Comment:
Probably worth defining field `type` values in an `enum`

Reply:
Done in 6b4b1f4.